### ISSUE TITLE
F821 Undefined name for variables loaded with notebook %store magic

### ISSUE THREAD

**Issue by athewsey** [astral-sh/ruff#25807]
Hi folks - Raising this as a blank issue as I think it's distinct from the related #9648 but not sure whether it's a bug or a feature request...

I'm using ruff to lint a notebook that loads variables from [IPython's %store magic extension](https://ipython.readthedocs.io/en/stable/config/extensions/storemagic.html) - like this:

```python
# Load specific variable saved from a previous, separate notebook:
%store -r my_cool_variable

# ...Then use it in this notebook:
print(my_cool_variable)
```

This currently (at ruff v0.15.16) gives error `F821 Undefined name 'my_cool_variable'` - which I would like to resolve if possible.

Would it be possible to detect and support this usage? Or is this a bug that we think should already work?

In the interim, could anybody suggest the best way(s) to suppress the error? Attaching `# noqa: F821` to each usage is not ideal in my case, as the variable is used many times in the notebook after the initial load.

### Additional context

Technically, `%store -r non_existent_variable` prints a warning message but does not raise an error - so it is true that a variable name *could* still be undefined after running this magic command... But I'd argue that from a linter's perspective we could just assume the name exists and has Unknown type - as a sensible default?

**Comment by ntBre** [comment]
I think we've been a bit hesitant to try parsing these magic directives in the past (I also found https://github.com/astral-sh/ruff/issues/10454 as a somewhat related case). Codex suggested the rather "creative" use of [`builtins`](https://docs.astral.sh/ruff/settings/#builtins) to suppress the error, assuming you don't want to use a file-level `ruff: noqa: F821` comment, which could conceal other undefined names in the same file.

**Comment by athewsey** [comment]
Yeah, I appreciate there are limits to how much this project would want to replicate magic directive logic...

If I understand right, the `builtins` would also need to be at the overall project level? So risks false negatives if that variable name is ever used before assignment in another file, too. It might be good if there's file-level builtins control?

For now I've implemented a workaround by following every %store retrieval with a noqa'd self-assignment, as shown below:

```python
%store -r my_cool_variable
my_cool_variable = my_cool_variable  # noqa: F821
```

Not sure I like it much, but it keeps the suppression tightly framed I guess 🤷‍♂️

**Comment by MichaReiser** [comment]
The easiest workaround is to default initialize the variable

```py
my_cool_variable: int  = 0
# Load specific variable saved from a previous, separate notebook:
%store -r my_cool_variable

# ...Then use it in this notebook:
print(my_cool_variable)
```

Ruff's behavior here is maybe somewhat pedantic but can be justified. E.g running the above code when the variable has never been stored fails with:

```
NameError                                 Traceback (most recent call last)
Cell In[1], [line 6](vscode-notebook-cell:?execution_count=1&line=6)
      3 get_ipython().run_line_magic('store', '-r my_cool_variable')
      5 # ...Then use it in this notebook:
----> [6](vscode-notebook-cell:?execution_count=1&line=6) print(my_cool_variable)

NameError: name 'my_cool_variable' is not defined
```
